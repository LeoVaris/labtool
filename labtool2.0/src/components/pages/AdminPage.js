import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Button, Loader, Label, Table, Popup } from 'semantic-ui-react'
import { Link } from 'react-router-dom'
import { Redirect } from 'react-router'
import { HorizontalScrollable } from '../HorizontalScrollable'
import { formatCourseName } from '../../util/format'
import { getAllCI } from '../../services/courseInstance'
import { resetLoading } from '../../reducers/loadingReducer'
import { getAllUsers, updateOtherUser } from '../../services/user'
import { clearNotifications } from '../../reducers/notificationReducer'
import { usePersistedState } from '../../hooks/persistedState'
import { sortUsersBySysopLastname } from '../../util/sort'

const AdminCourseList = ({ courses }) => {
  if (!Array.isArray(courses)) {
    return <div />
  }

  return (
    <HorizontalScrollable>
      <Table singleLine color="yellow" style={{ overflowX: 'visible' }}>
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell colSpan="1">Course ID</Table.HeaderCell>
            <Table.HeaderCell colSpan="1">Course name</Table.HeaderCell>
            <Table.HeaderCell colSpan="1">Course start date</Table.HeaderCell>
            <Table.HeaderCell colSpan="2"> </Table.HeaderCell>
          </Table.Row>
        </Table.Header>

        <Table.Body>
          {courses.map(instance => (
            <Table.Row key={instance.id}>
              <Table.Cell>{instance.shorterId} </Table.Cell>
              <Table.Cell>
                <strong>{formatCourseName(instance.name, instance.ohid, instance.start)}</strong>
              </Table.Cell>

              <Table.Cell> {instance.europeanStart} </Table.Cell>
              <Table.Cell textAlign="center">
                <Popup
                  trigger={
                    <Button
                      circular
                      size="tiny"
                      icon={{ name: 'edit', size: 'large', color: 'orange' }}
                      as={Link}
                      to={{
                        pathname: `/labtool/ModifyCourseInstanceStaff/${instance.ohid}`,
                        state: { fromAdmin: true }
                      }}
                    />
                  }
                  position="top right"
                  content="Manage instructors for this course"
                />
              </Table.Cell>
            </Table.Row>
          ))}
        </Table.Body>
      </Table>
    </HorizontalScrollable>
  )
}

AdminCourseList.propTypes = {
  courses: PropTypes.oneOfType([PropTypes.array, PropTypes.object])
}

const AdminUserList = ({ me, users, updateOtherUser }) => {
  const setAdmin = (userId, name, newAdmin) => () => {
    const message = `Are you sure you want to ${newAdmin ? 'grant admin rights to' : 'remove admin rights from'} ${name}?`
    if (window.confirm(message)) {
      const selfInflicted = userId === me.user.id
      if (selfInflicted && !newAdmin && !window.confirm('You probably do NOT want to remove admin rights from yourself. Are you really sure?')) {
        return
      }
      updateOtherUser({ id: userId, sysop: newAdmin })
    }
  }
  const sortedUsers = sortUsersBySysopLastname(users)

  return (
    <Table singleLine color="yellow" fixed>
      <Table.Header>
        <Table.Row>
          <Table.HeaderCell>Name</Table.HeaderCell>
          <Table.HeaderCell>Status</Table.HeaderCell>
        </Table.Row>
      </Table.Header>
      <Table.Body>
        {sortedUsers.map(user => (
          <Table.Row key={user.id}>
            <Table.Cell>
              {user.firsts} {user.lastname}
              <br />
              <a href={`mailto:${user.email}`}>{user.email}</a>
            </Table.Cell>
            <Table.Cell>
              {user.teacher && (
                <Label color="orange" horizontal>
                  Teacher
                </Label>
              )}
              {user.sysop ? (
                <div>
                  <Label color="orange" horizontal>
                    Admin
                  </Label>
                  <Button onClick={setAdmin(user.id, `${user.firsts} ${user.lastname}`, false)} size="tiny" color="green">
                    Remove admin
                  </Button>
                </div>
              ) : (
                <div>
                  <Label horizontal>Not admin</Label>
                  <Button onClick={setAdmin(user.id, `${user.firsts} ${user.lastname}`, true)} size="tiny" color="green">
                    Make admin
                  </Button>
                </div>
              )}
            </Table.Cell>
          </Table.Row>
        ))}
      </Table.Body>
    </Table>
  )
}

AdminUserList.propTypes = {
  me: PropTypes.object.isRequired,
  users: PropTypes.array.isRequired,
  updateOtherUser: PropTypes.func.isRequired
}

export const AdminPage = props => {
  const state = usePersistedState('AdminPage', {
    coursesList: false,
    usersList: false
  })

  useEffect(() => {
    props.clearNotifications()
  }, [])

  useEffect(() => {
    if (state.coursesList) {
      props.resetLoading()
      props.getAllCI()
    }
  }, [state.coursesList])

  useEffect(() => {
    if (state.usersList) {
      props.resetLoading()
      props.getAllUsers()
    }
  }, [state.usersList])

  if (!props.user.user.sysop) {
    return <Redirect to={`/labtool`} />
  } else if (props.loading.loading) {
    return <Loader active />
  }

  return (
    <div className="AdminPage">
      <h2>Welcome to the admin interface.</h2>
      <Button compact className={`ui button`} onClick={() => (state.coursesList = !state.coursesList)}>
        {state.coursesList ? 'Hide courses' : 'Manage courses'}
      </Button>
      <br />
      <br />
      {state.coursesList && <AdminCourseList courses={props.courseInstance} />}
      <br />
      <hr />
      <br />
      <Button compact className={`ui button`} onClick={() => (state.usersList = !state.usersList)}>
        {state.usersList ? 'Hide users' : 'Manage users'}
      </Button>
      <br />
      {state.usersList && <AdminUserList me={props.user} users={props.users} updateOtherUser={props.updateOtherUser} />}
    </div>
  )
}

const mapStateToProps = state => {
  return {
    user: state.user,
    courseInstance: state.courseInstance,
    users: state.users,
    loading: state.loading
  }
}

const mapDispatchToProps = {
  getAllCI,
  resetLoading,
  getAllUsers,
  updateOtherUser,
  clearNotifications
}

AdminPage.propTypes = {
  user: PropTypes.object.isRequired,
  courseInstance: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  users: PropTypes.array.isRequired,
  loading: PropTypes.object.isRequired,

  getAllCI: PropTypes.func.isRequired,
  resetLoading: PropTypes.func.isRequired,
  getAllUsers: PropTypes.func.isRequired,
  updateOtherUser: PropTypes.func.isRequired,
  clearNotifications: PropTypes.func.isRequired
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AdminPage)
