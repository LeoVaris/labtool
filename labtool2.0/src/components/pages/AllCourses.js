import React, { Component } from 'react'
import { getAll } from '../../services/courseInstance'
import { Link } from 'react-router-dom'

class AllCourses extends Component {
  constructor(props) {
    super(props)
    this.state = {
      courses: [],
    }
  }

  componentDidMount() {
    getAll().then(courses =>
      this.setState({ courses })
    )
  }

  render() {
    return (
      <div>
        <ul>
          {
            this.state.courses.map(c => (
              <li key={c.id}>
                <Link to={`/courses/${c.id}`}>{c.name}</Link>
              </li>
            ))
          }
        </ul>
      </div>
    )
  }
}



export default AllCourses