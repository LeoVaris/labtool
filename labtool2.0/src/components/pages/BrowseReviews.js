import React, { Component } from 'react'
import { Button, Card, Accordion, Icon, Form, Comment, Input, Popup, Loader, Label } from 'semantic-ui-react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { createOneComment } from '../../services/comment'
import { getOneCI, coursePageInformation } from '../../services/courseInstance'
import { gradeCodeReview } from '../../services/codeReview'
import ReactMarkdown from 'react-markdown'
import { sendEmail } from '../../services/email'
import { resetLoading } from '../../reducers/loadingReducer'
import { trimDate, createCourseIdWithYearAndTerm } from '../../util/format'
import { getCoursesByStudentId } from '../../services/studentinstances'

import BackButton from '../BackButton'
import { FormMarkdownTextArea } from '../MarkdownTextArea'

/**
 * Maps all comments from a single instance from coursePage reducer
 */
export class BrowseReviews extends Component {
  constructor(props) {
    super(props)
    this.state = {
      openWeeks: {},
      initialLoading: props.initialLoading !== undefined ? this.props.initialLoading : true
    }
  }

  componentWillMount = async () => {
    await this.props.resetLoading()
    this.props.getOneCI(this.props.courseId)
    this.props.coursePageInformation(this.props.courseId)
    this.props.getCoursesByStudentId(Number(this.props.studentInstance))
  }

  componentDidMount() {
    if (!this.props.loading.loading && !this.state.openWeeks[this.props.selectedInstance.currentWeek - 1]) {
      this.setState({ openWeeks: { [this.props.selectedInstance.currentWeek - 1]: true } })
    }
  }

  componentDidUpdate() {
    if (!this.props.loading.loading && this.state.initialLoading) {
      this.setState({ initialLoading: false })
    }
  }

  handleClick = (e, titleProps) => {
    const { index } = titleProps
    const { openWeeks } = this.state
    const numberOfOpen = Object.values(openWeeks).filter(x => x).length

    if (numberOfOpen === 0) {
      // if no reviews open, open the clicked tab
      this.setState({ openWeeks: { [index]: true } })
    } else if (numberOfOpen === 1) {
      // if one open, open the clicked tab and close everything else,
      // or close everything if we clicked the open one
      this.setState({ openWeeks: { [index]: !openWeeks[index] } })
    } else {
      // if multiple open, simply toggle open/close
      this.setState({ openWeeks: { ...openWeeks, [index]: !openWeeks[index] } })
    }
  }

  getMaximumIndexForStudent = student => {
    // how many reviews will there be?
    return this.props.selectedInstance.weekAmount + student.codeReviews.length + 1
  }

  hasAllReviewsOpen = student => {
    return Object.values(this.state.openWeeks).filter(x => x).length == this.getMaximumIndexForStudent(student)
  }

  handleClickShowAll = student => () => {
    const openWeeks = {}
    const maximumIndex = this.getMaximumIndexForStudent(student)
    for (let i = 0; i < maximumIndex; i++) {
      openWeeks[i] = true
    }
    this.setState({ openWeeks })
  }

  handleClickHideAll = () => {
    this.setState({ openWeeks: {} })
  }

  handleSubmit = async e => {
    e.preventDefault()
    const content = {
      hidden: e.target.hidden.checked,
      comment: e.target.content.value,
      week: parseInt(e.target.name, 10)
    }
    document.getElementById(e.target.name).reset()
    try {
      await this.props.createOneComment(content)
    } catch (error) {
      console.log(error)
    }
  }

  sortCommentsByDate = comments => {
    return comments.sort((a, b) => {
      return new Date(a.createdAt) - new Date(b.createdAt)
    })
  }

  gradeCodeReview = (reviewNumber, studentInstanceId) => async e => {
    e.preventDefault()
    const data = {
      reviewNumber,
      studentInstanceId: Number(studentInstanceId),
      points: Number(e.target.points.value)
    }
    this.props.gradeCodeReview(data)
  }

  sendCommentEmail = commentId => async () => {
    this.props.sendEmail({
      commentId,
      role: 'teacher'
    })
  }

  sendWeekEmail = weekId => async () => {
    this.props.sendEmail({
      weekId,
      role: 'teacher'
    })
  }

  isTeacher = () => {
    return this.props.courseData.role === 'teacher'
  }

  //get student's other participations in the same course
  renderStudentPreviousParticipation = () => {
    const previousParticipations = this.props.studentInstanceToBeReviewed.filter(
      courseInstance => courseInstance.ohid.includes(this.props.courseId.substring(0, 8)) && courseInstance.ohid !== this.props.courseId
    )
    if (previousParticipations.length === 0) {
      return <p className="noPrevious">First time to participate the course</p>
    }
    return (
      <div className="hasPrevious">
        <p className style={{ color: 'red' }}>
          Has other participations
        </p>
        {previousParticipations.map(participation => <p key={participation.id}>{createCourseIdWithYearAndTerm(participation.ohid, participation.start)}</p>)}
      </div>
    )
  }

  renderStudentCard = student => (
    <Card key={student.id} fluid color="yellow" className="studentCard">
      <Card.Content>
        <h2>
          {student.User.firsts} {student.User.lastname} ({student.User.studentNumber})
        </h2>
        <h3>
          <a href={`mailto:${student.User.email}`}>{student.User.email}</a>
        </h3>
        <h3>
          {student.projectName}{' '}
          <a href={student.github} target="_blank" rel="noopener noreferrer">
            {student.github}
          </a>
        </h3>
        {this.renderStudentPreviousParticipation()}
      </Card.Content>
    </Card>
  )

  renderComment = isFinalWeek => comment =>
    comment.hidden ? (
      <Comment disabled>
        <Comment.Content>
          <Comment.Metadata>
            <div>Hidden</div>
          </Comment.Metadata>
          <Comment.Author>{comment.from}</Comment.Author>
          <Comment.Text>
            {' '}
            <ReactMarkdown>{comment.comment}</ReactMarkdown>{' '}
          </Comment.Text>
          <Comment.Metadata>
            <div>{trimDate(comment.createdAt)}</div>
          </Comment.Metadata>
          <div> </div>
        </Comment.Content>
      </Comment>
    ) : (
      <Comment key={comment.id}>
        <Comment.Author>{comment.from}</Comment.Author>
        <Comment.Text>
          {' '}
          <ReactMarkdown>{comment.comment}</ReactMarkdown>{' '}
        </Comment.Text>
        <Comment.Metadata>
          <div>{trimDate(comment.createdAt)}</div>
        </Comment.Metadata>
        {!isFinalWeek ? (
          <div>
            <div> </div>
            {/* This hack compares user's name to comment.from and hides the email notification button when they don't match. */}
            {comment.from.includes(this.props.user.user.firsts) && comment.from.includes(this.props.user.user.lastname) ? (
              comment.notified ? (
                <Label>
                  Notified <Icon name="check" color="green" />
                </Label>
              ) : (
                <Button type="button" onClick={this.sendCommentEmail(comment.id)} size="small">
                  Send email notification
                </Button>
              )
            ) : (
              <div />
            )}
          </div>
        ) : (
          <div />
        )}
      </Comment>
    )

  renderWeek = (i, week, studentInstance, isFinalWeek) => {
    const { openWeeks } = this.state
    const reviewIndex = isFinalWeek ? this.props.selectedInstance.weekAmount + 1 : i + 1

    if (week) {
      return (
        <Accordion fluid styled>
          <Accordion.Title active={openWeeks[i]} index={i} onClick={this.handleClick}>
            <Icon name="dropdown" /> {isFinalWeek ? 'Final Review' : `Week ${i + 1}`}, points {week.points}
          </Accordion.Title>
          <Accordion.Content active={openWeeks[i]}>
            <Card fluid color="yellow">
              <Card.Content>
                <h4> Points {week.points} </h4>
                <h4> Feedback </h4>
                <ReactMarkdown>{week.feedback}</ReactMarkdown>{' '}
                {this.isTeacher() && week.instructorNotes ? (
                  <div>
                    <br />
                    <h4>Review notes for instructors </h4>
                    <ReactMarkdown>{week.instructorNotes}</ReactMarkdown>{' '}
                  </div>
                ) : (
                  <span />
                )}
              </Card.Content>
              {!isFinalWeek ? (
                <Card.Content style={{ paddingBottom: '5px' }}>
                  {week.notified ? (
                    <Label>
                      Notified <Icon name="check" color="green" />
                    </Label>
                  ) : (
                    <Button type="button" onClick={this.sendWeekEmail(week.id)} size="small">
                      Send email notification
                    </Button>
                  )}
                </Card.Content>
              ) : (
                <span />
              )}
            </Card>
            <h4> Comments </h4>
            <Comment.Group>{week ? this.sortCommentsByDate(week.comments).map(this.renderComment(isFinalWeek)) : <h4> No comments </h4>}</Comment.Group>
            <Form reply onSubmit={this.handleSubmit} name={week.id} id={week.id}>
              <FormMarkdownTextArea name="content" placeholder="Your comment..." defaultValue="" />
              <Form.Checkbox label="Add comment for instructors only" name="hidden" />
              <Button content="Add Reply" labelPosition="left" icon="edit" primary />
            </Form>
            <h3>Review</h3>
            <Link to={`/labtool/reviewstudent/${this.props.selectedInstance.ohid}/${studentInstance}/${reviewIndex}`}>
              <Popup trigger={<Button circular color="orange" size="tiny" icon={{ name: 'edit', color: 'black', size: 'large' }} />} content={isFinalWeek ? 'Edit final review' : 'Edit review'} />
            </Link>
          </Accordion.Content>
        </Accordion>
      )
    } else {
      return (
        <Accordion fluid styled>
          <Accordion.Title active={openWeeks[i]} index={i} onClick={this.handleClick}>
            <Icon name="dropdown" /> {isFinalWeek ? 'Final Review' : `Week ${i + 1}`}{' '}
          </Accordion.Title>
          <Accordion.Content active={openWeeks[i]}>
            <h4> Not Graded </h4>
            <h4> No comments </h4>
            <Link to={`/labtool/reviewstudent/${this.props.selectedInstance.ohid}/${studentInstance}/${reviewIndex}`}>
              <Popup trigger={<Button circular color="orange" size="tiny" icon={{ name: 'edit', color: 'black', size: 'large' }} />} content={isFinalWeek ? 'Give Final Review' : 'Review week'} />
            </Link>
          </Accordion.Content>
        </Accordion>
      )
    }
  }

  renderCodeReview = (i, cr, studentInstance) => {
    const { openWeeks } = this.state

    return (
      <Accordion fluid styled>
        {' '}
        <Accordion.Title active={openWeeks[i]} index={i} onClick={this.handleClick}>
          <Icon name="dropdown" /> Code Review {cr.reviewNumber} {cr.points !== null ? ', points ' + cr.points : ''}
        </Accordion.Title>
        <Accordion.Content active={openWeeks[i]}>
          <p>
            <strong>Project to review:</strong> {this.props.courseData.data.find(data => data.id === cr.toReview).projectName} <br />
            <strong>GitHub:</strong>{' '}
            <a href={this.props.courseData.data.find(data => data.id === cr.toReview).github} target="_blank" rel="noopener noreferrer">
              {this.props.courseData.data.find(data => data.id === cr.toReview).github}
            </a>
          </p>
          <strong>Code review:</strong>{' '}
          {cr.linkToReview ? (
            <a href={cr.linkToReview} target="_blank" rel="noopener noreferrer">
              {cr.linkToReview}
            </a>
          ) : (
            'No review linked yet'
          )}
          {cr.points !== null ? <h4>{cr.points} points</h4> : <h4>Not graded yet</h4>}
          <Form onSubmit={this.gradeCodeReview(cr.reviewNumber, studentInstance)}>
            <label>Points </label>
            <Input name="points" defaultValue={cr.points ? cr.points : ''} type="number" step="0.01" style={{ width: '100px' }} />
            <Input type="submit" value="Grade" />
          </Form>
        </Accordion.Content>
      </Accordion>
    )
  }

  render() {
    if (this.state.initialLoading) {
      return <Loader active />
    }

    const createHeaders = (studhead, studentInstance) => {
      let headers = []
      studhead.data.map(student => {
        // studentInstance is id of student. Type: String
        // Tämä pitää myös korjata.
        if (student.id === Number(studentInstance)) {
          headers.push(this.renderStudentCard(student))
          headers.push(
            <span>
              {this.hasAllReviewsOpen(student) ? (
                <Button type="button" onClick={this.handleClickHideAll} size="small">
                  Hide all reviews
                </Button>
              ) : (
                <Button type="button" onClick={this.handleClickShowAll(student)} size="small">
                  Show all reviews
                </Button>
              )}
              <br />
            </span>
          )
          let i = 0
          let ii = 0
          for (; i < this.props.selectedInstance.weekAmount; i++) {
            const weeks = student.weeks.find(week => week.weekNumber === i + 1)
            headers.push(this.renderWeek(i, weeks, studentInstance, false))
          }
          student.codeReviews
            .sort((a, b) => {
              return a.reviewNumber - b.reviewNumber
            })
            .forEach(cr => {
              headers.push(this.renderCodeReview(i + ii + 1, cr, studentInstance))
              ii++
            })
          if (this.props.selectedInstance.finalReview) {
            const finalWeek = student.weeks.find(week => week.weekNumber === this.props.selectedInstance.weekAmount + 1)
            headers.push(this.renderWeek(i + ii, finalWeek, studentInstance, true))
          }
        }
        return student
      })
      return headers.map((header, index) => React.cloneElement(header, { key: index }))
    }

    return (
      <div className="BrowseReviews" style={{ overflowX: 'auto' }}>
        <Loader active={this.props.loading.loading} />
        {this.isTeacher() ? (
          <div>
            <BackButton preset="coursePage" />
            <Link to={`/labtool/courses/${this.props.selectedInstance.ohid}`} style={{ textAlign: 'center' }}>
              <h2> {this.props.selectedInstance.name} </h2>
            </Link>
            {createHeaders(this.props.courseData, this.props.studentInstance)}
          </div>
        ) : (
          <p />
        )}
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    ...ownProps,
    user: state.user,
    selectedInstance: state.selectedInstance,
    courseData: state.coursePage,
    studentInstanceToBeReviewed: state.studentInstance,
    loading: state.loading
  }
}

const mapDispatchToProps = {
  createOneComment,
  getOneCI,
  coursePageInformation,
  gradeCodeReview,
  sendEmail,
  getCoursesByStudentId: getCoursesByStudentId,
  resetLoading
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BrowseReviews)
