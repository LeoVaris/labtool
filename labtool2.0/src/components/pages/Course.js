import React, { Component } from 'react'
import { getOne } from '../../services/courseInstance'



class Course extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      course: undefined,
    }
  }

  componentDidMount() {
    getOne(parseInt(this.props.match.params.number, 10)).then(
      course => this.setState({ course })
    )
  }

  render() {
    if (!this.state.course) {
      return <div>course not found!!!</div>
    }

    return (
      <div>
        <h1>{this.state.course.name} </h1>
      </div>
    )
  }
}

export default Course