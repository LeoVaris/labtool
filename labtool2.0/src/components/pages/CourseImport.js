import React, { Component } from 'react'
import { Button, Container, Header, Table, Loader, Form } from 'semantic-ui-react'
import { connect } from 'react-redux'
import { Link, Redirect } from 'react-router-dom'
import { getIsAllowedToImport, getImportableCourses, importCourses } from '../../services/courseImport'
import { resetLoading, addRedirectHook } from '../../reducers/loadingReducer'
import { HorizontalScrollable } from '../HorizontalScrollable'

/**
 *  Show all the courses in a single list.
 */
export class CourseImport extends Component {
  componentWillMount = async () => {
    await this.props.resetLoading()
    this.props.getIsAllowedToImport()
    this.props.getImportableCourses()
  }

  handleSubmit = async e => {
    e.preventDefault()
    const data = this.props.importable
    if (data) {
      const courses = data.filter(entry => e.target['import' + entry.hid] && e.target['import' + entry.hid].checked)
      this.props.addRedirectHook({
        hook: 'COURSE_IMPORT_DO_IMPORT_'
      })
      await this.props.importCourses({ courses })
    }
  }

  render() {
    return (
      <div className="CourseImport">
        <Container>
          <Header as="h2">Import courses</Header>
          {this.props.loading.loading || !Array.isArray(this.props.importable) || !this.props.canImport ? (
            <Loader active />
          ) : this.props.loading.redirect ? (
            <Redirect to="/labtool/courses" />
          ) : this.props.importable.length ? (
            <Form onSubmit={this.handleSubmit}>
              <HorizontalScrollable>
                <Table singleLine color="yellow" style={{ overflowX: 'visible' }}>
                  <Table.Header>
                    <Table.Row>
                      <Table.HeaderCell colSpan="1" key={-1}>
                        Import?
                      </Table.HeaderCell>
                      <Table.HeaderCell colSpan="1">Course id</Table.HeaderCell>
                      <Table.HeaderCell colSpan="1">Course name</Table.HeaderCell>
                      <Table.HeaderCell colSpan="1">Course start date</Table.HeaderCell>
                      <Table.HeaderCell colSpan="1">Course end date</Table.HeaderCell>
                    </Table.Row>
                  </Table.Header>

                  <Table.Body>
                    {this.props.importable.map(instance => (
                      <Table.Row key={instance.hid}>
                        <Table.Cell>
                          <Form.Checkbox name={'import' + instance.hid} />
                        </Table.Cell>
                        <Table.Cell>{instance.shorterId}</Table.Cell>
                        <Table.Cell>{instance.cname}</Table.Cell>
                        <Table.Cell>{instance.europeanStart}</Table.Cell>
                        <Table.Cell>{instance.europeanEnd}</Table.Cell>
                      </Table.Row>
                    ))}
                  </Table.Body>
                </Table>
              </HorizontalScrollable>

              <br />
              <br />

              <Button type="submit" className="ui green button" content="Import" />

              <br />
              <br />

              <Link to="/labtool/courses">
                <Button className="ui button" type="cancel">
                  Cancel
                </Button>
              </Link>
            </Form>
          ) : (
            <div>
              <h4>There are no courses in Kurki to import</h4>

              <Link to="/labtool/courses">
                <Button className="ui button" type="cancel">
                  Back
                </Button>
              </Link>
            </div>
          )}
        </Container>
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    loading: state.loading,
    canImport: state.courseImport.canImport,
    importable: state.courseImport.importable
  }
}

const mapDispatchToProps = {
  resetLoading,
  getIsAllowedToImport,
  getImportableCourses,
  importCourses,
  addRedirectHook
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CourseImport)
