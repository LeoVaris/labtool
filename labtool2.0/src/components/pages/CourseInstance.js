import React from 'react'

const CourseInstance = ({ instance, handleFieldChange }) => {
  return(
    <li style={{ textAlignVertical: 'center', textAlign: 'center' }}> {instance.name} <button name={instance.name} onClick={handleFieldChange} value={instance.id}>Register</button>
    </li>
  )
}

export default CourseInstance