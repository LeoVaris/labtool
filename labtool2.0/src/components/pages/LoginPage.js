import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { login } from '../../services/login'
import { resetLoading, forceSetLoading } from '../../reducers/loadingReducer'
import { Form, Input, Button, Grid, Loader } from 'semantic-ui-react'

/**
 *  The page used to login
 */

export const LoginPage = props => {
  useEffect(() => {
    // run on component mount
    props.resetLoading()
    props.forceSetLoading({
      value: false
    })
  }, [])

  const handleSubmit = async e => {
    e.preventDefault()

    const content = {
      username: e.target.username.value,
      password: e.target.password.value
    }
    await props.login(content)
  }

  return (
    <div className="LoginPage">
      <Loader active={props.loading.loading} inline="centered" />

      <Grid>
        <Grid.Row centered>
          <h3>Enter your University of Helsinki username and password</h3>
        </Grid.Row>
      </Grid>

      <Grid>
        <Grid.Row centered>
          <Form onSubmit={handleSubmit}>
            <Form.Group inline>
              <label style={{ width: '75px' }}>Username</label>
              <Input type="text" name="username" icon="user" required="true" iconPosition="left" style={{ minWidth: '25em' }} placeholder="Your AD-username" className="form-control1" />
            </Form.Group>

            <Form.Group inline>
              <label style={{ width: '75px' }}>Password</label>
              <Input type="password" name="password" required="true" icon="lock" iconPosition="left" style={{ minWidth: '25em' }} placeholder="Your password" className="form-control2" />
            </Form.Group>

            <Form.Group>
              <Button type="submit" color="blue">
                Login
              </Button>
            </Form.Group>
          </Form>
        </Grid.Row>
      </Grid>
    </div>
  )
}
const mapStateToProps = state => {
  return {
    user: state.user,
    loading: state.loading
  }
}

const mapDispatchToProps = {
  login,
  resetLoading,
  forceSetLoading
}

LoginPage.propTypes = {
  user: PropTypes.object.isRequired,
  loading: PropTypes.object.isRequired,

  login: PropTypes.func.isRequired,
  resetLoading: PropTypes.func.isRequired,
  forceSetLoading: PropTypes.func.isRequired
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginPage)
