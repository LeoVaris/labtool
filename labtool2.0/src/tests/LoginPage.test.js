import React from 'react'
import { LoginPage } from '../components/pages/LoginPage'
import { shallow } from 'enzyme'

describe('<Login />', () => {
  let wrapper

  const loading = {
    loading: false,
    loadingHooks: [],
    redirect: false,
    redirectHooks: [],
    redirectFailure: false
  }

  const mockFn = jest.fn()

  beforeEach(() => {
    wrapper = shallow(<LoginPage user={{}} loading={loading} resetLoading={mockFn} forceSetLoading={mockFn} login={mockFn} />)
  })

  describe('LoginPage Component', () => {
    it('is ok', () => {
      true
    })

    it('should render without throwing an error', () => {
      expect(wrapper.find('.LoginPage').length).toEqual(1)
    })

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('renders a username input', () => {
      expect(wrapper.find('.form-control1').length).toEqual(1)
    })

    it('renders a password input', () => {
      expect(wrapper.find('.form-control2').length).toEqual(1)
    })
  })
})
