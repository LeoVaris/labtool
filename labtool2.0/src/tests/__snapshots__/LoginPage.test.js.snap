// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> LoginPage Component should render correctly 1`] = `
<div
  className="LoginPage"
>
  <Loader
    active={false}
    inline="centered"
  />
  <Grid>
    <GridRow
      centered={true}
    >
      <h3>
        Enter your University of Helsinki username and password
      </h3>
    </GridRow>
  </Grid>
  <Grid>
    <GridRow
      centered={true}
    >
      <Form
        as="form"
        onSubmit={[Function]}
      >
        <FormGroup
          inline={true}
        >
          <label
            style={
              Object {
                "width": "75px",
              }
            }
          >
            Username
          </label>
          <Input
            className="form-control1"
            icon="user"
            iconPosition="left"
            name="username"
            placeholder="Your AD-username"
            required="true"
            style={
              Object {
                "minWidth": "25em",
              }
            }
            type="text"
          />
        </FormGroup>
        <FormGroup
          inline={true}
        >
          <label
            style={
              Object {
                "width": "75px",
              }
            }
          >
            Password
          </label>
          <Input
            className="form-control2"
            icon="lock"
            iconPosition="left"
            name="password"
            placeholder="Your password"
            required="true"
            style={
              Object {
                "minWidth": "25em",
              }
            }
            type="password"
          />
        </FormGroup>
        <FormGroup>
          <Button
            as="button"
            color="blue"
            role="button"
            type="submit"
          >
            Login
          </Button>
        </FormGroup>
      </Form>
    </GridRow>
  </Grid>
</div>
`;
